### ISSUE TITLE
Option to reduce verbosity of Syncer browser 

### ISSUE THREAD

**Issue by danielo515** [Jermolene/TiddlyWiki5#2681]
Logging is important to develop an analyze the behavior of a program, but having too much can be as bad as not having any logging at all. The messages of the syncer browser on the console usually annoys me when I try to debug my own plugins. It adds a lot of noise and usually puts me on a hurry to analyze what it's happening before it pukes again all their logs messages. 

Having an option to reduce the verbosity of what is being logged (for example, only when fetching starts and ends) will be very welcome. Ideally you should be able to set the verbosity level directly from the console calling a method of the syncer itself. 

Regards

**Comment by Jermolene** [comment]
Hi @danielo515 adding the ability to filter the log messages would certainly be useful, if you'd like to work on a pull request?

The code for the logger class is here:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/utils/logger.js

It is used in several places (search for `$tw.utils.Logger`), for example:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/syncer.js#L25
https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/syncer.js#L160

It would be relatively easy to filter messages based on the component name.

**Comment by danielo515** [comment]
Hello @Jermolene ,

I was talking about just the syncer module. I suppose that it is better to implement the functionality on the root logger and then control the verbosity on the different modules that uses them. 

But in order to control verbosity, I think we should implement a couple more of methods. I think the standard ones are fine `info`  and `log` for useful information and in case of low verbosity  and `debug` and  `trace` for verbose logging output.  Maybe `warn` and `error` could make sense too, but I don't see them as very relevant here.

Regards

**Comment by Jermolene** [comment]
> I was talking about just the syncer module. I suppose that it is better to implement the functionality on the root logger and then control the verbosity on the different modules that uses them.

Yes. The primary purpose of having a centralised logging class is so that we can implement logging features consistently across different subsystems. Filtering is certainly one of the basic things that logging libraries usually provide.

> But in order to control verbosity, I think we should implement a couple more of methods. I think the standard ones are fine info and log for useful information and in case of low verbosity and debug and trace for verbose logging output. Maybe warn and error could make sense too, but I don't see them as very relevant here.

The distinction between the current `info` and `log` methods is that the former is intended for developer-focussed messages, and the latter presents messages that are visible to the user.

What would be the distinction between `debug` and `trace`?

**Comment by danielo515** [comment]
For me `debug` and `trace` are more or less the same concept but considered two different verbosity levels.

- Trace - Only when I would be "tracing" the code and trying to find one part of a function specifically.
- Debug - Information that is diagnostically helpful to people more than just developers (IT, sysadmins, etc.)

To summarize, on debug mode we will see debug output while on verbose debug we will se both debug and trace.

Regards

**Comment by Jermolene** [comment]
Hi @danielo515 it sounds like you've got two concepts there: the different logging methods (log vs. info vs. trace vs. debug), and the current logging level (normal vs. debug vs. verbose debug). Is that right?

**Comment by danielo515** [comment]
Personally I see them quite coupled. When I'm using a logger I just do `Logger.trace(...)` when I'm logging something that I only what to ve visible on verbose debug mode, `Logger.debug(...)` when I want logging information that will be visible on normal debug mode and the rest for all the cases. How else would you distinguish the messages that should not be displayed on non-verbose debug modes?

**Comment by Jermolene** [comment]
Yes, so we have different methods that do different types of logging. But how does the user control which logging messages are visible? In other words, how do they select debug vs. non-debug, verbose vs. non-verbose.

**Comment by danielo515** [comment]
Oh, that is a totally different point.
In my oppinion, the default state should be non-debug level. I think that putting efforts on creating any kind of UI for something that will force you to open dev tools to see the result is a waste of time. Ideally I would like to be able to specify the debug level of each module. Probably a global method on `$tw` object would be enough. Something like `$tw.debug('module-name',{verbose:true})` while the second object is optional, and the `module-name` can be an asterisk to debug everything.

What do you think?

**Comment by Jermolene** [comment]
> Oh, that is a totally different point.

I'm not sure that controlling the log output is a different point. Surely it is integral to your OP asking for a way to suppress (some of) the logging of the syncer module?

> In my oppinion, the default state should be non-debug level

Can you explain why?

>  I think that putting efforts on creating any kind of UI for something that will force you to open dev tools to see the result is a waste of time

OK. Is anyone suggesting that there should be a UI?

>  Ideally I would like to be able to specify the debug level of each module. Probably a global method on $tw object would be enough. Something like $tw.debug('module-name',{verbose:true}) while the second object is optional, and the module-name can be an asterisk to debug everything.

That seems pretty unfriendly. To persistently change the current logging level the user would have to create a JS module with that command in it.

**Comment by danielo515** [comment]
> I'm not sure that controlling the log output is a different point. Surely it is integral to your OP asking for a way to suppress (some of) the logging of the syncer module?

Sorry, I meant that it was a different from what I said on my previous message. It was not my intention to refer the entire conversation.

> Can you explain why?
Just feel it that way. Why do the non-dev users need the continuos logging messages on the console? I'm not sure if this has any performance impact.

> OK. Is anyone suggesting that there should be a UI?
I just took it for granted. My fault.

> That seems pretty unfriendly. To persistently change the current logging level the user would have to create a JS module with that command in it.

I'm open to any suggestion. Maybe overwriting some system/shadow tiddler ?
